java/java-analysis-impl/src/com/siyeh/ig/psiutils/FinalUtils.java

```diff
@@ -26,28 +26,44 @@ public static boolean canBeFinal(@NotNull PsiVariable variable) {
     if (variable instanceof PsiField && !HighlightControlFlowUtil.isFieldInitializedAfterObjectConstruction((PsiField)variable)) {
       return false;
     }
+    return checkIfElementViolatesFinality(variable);
+  }
+
+  private static boolean checkIfElementViolatesFinality(@NotNull PsiVariable variable) {
     PsiElement scope = variable instanceof PsiField
                        ? PsiUtil.getTopLevelClass(variable)
                        : PsiUtil.getVariableCodeBlock(variable, null);
     if (scope == null) return false;
     Map<PsiElement, Collection<ControlFlowUtil.VariableInfo>> finalVarProblems = new HashMap<>();
     Map<PsiElement, Collection<PsiReferenceExpression>> uninitializedVarProblems = new HashMap<>();
     PsiElementProcessor<PsiElement> elementDoesNotViolateFinality = e -> {
-      if (!(e instanceof PsiReferenceExpression ref)) return true;
-      if (!ref.isReferenceTo(variable)) return true;
-      HighlightInfo.Builder highlightInfo = HighlightControlFlowUtil
-        .checkVariableInitializedBeforeUsage(ref, variable, uninitializedVarProblems, variable.getContainingFile(), true);
-      if (highlightInfo != null) return false;
-      if (!PsiUtil.isAccessedForWriting(ref)) return true;
-      if (!LocalsOrMyInstanceFieldsControlFlowPolicy.isLocalOrMyInstanceReference(ref)) return false;
-      if (ControlFlowUtil.isVariableAssignedInLoop(ref, variable)) return false;
-      if (variable instanceof PsiField) {
-        if (PsiUtil.findEnclosingConstructorOrInitializer(ref) == null) return false;
-        PsiElement innerScope = HighlightControlFlowUtil.getElementVariableReferencedFrom(variable, ref);
-        if (innerScope != null && innerScope != ((PsiField)variable).getContainingClass()) return false;
-      }
-      return HighlightControlFlowUtil.checkFinalVariableMightAlreadyHaveBeenAssignedTo(variable, ref, finalVarProblems) == null;
+      return checkElementDoesNotViolateFinality(e, variable, uninitializedVarProblems, finalVarProblems);
     };
     return PsiTreeUtil.processElements(scope, elementDoesNotViolateFinality);
   }
+
+  /*
+  * this method is used in the Kotlin plugin J2K process
```

Comment:
Please write a javadoc concentrating on what this new public method does and what its parameters mean instead of where it's used. It's not so necessary to say where it's used. Thank you.

Reply:
updated! thank you!